### PR TITLE
Asynchronous Incoming External Message validation

### PR DESCRIPTION
**Pull request by arc-vmware**
The message validation for various message involves
many time consuming tasks. Such tasks leads to slower
validation of messages. If the entire process of
validation happens asynchronously then the throughput
of our system will improve. This PR is achieving this
by the following step:
If the message requires asynchronous validation
then the validation happens in a separate thread;
     After validation, the message is translated
     into an internal message and pushed into the
     internal message queue.
     The internal message is handled by the replica
Criteria for async message validation:
If for a message the validation require any time consuming
operation, like signature verification, then only such message's
validation will happen asynchronously.
List of messages chosen in this PR are as below:
1) CheckpointMsg
2) ClientRequestMsg
3) PrePrepareMsg (This message is handled separately)
4) ReplicaRestartReadyMsg
5) ReplicasRestartReadyProofMsg
6) ViewChangeMsg
Adding existing message for Async validation is simple:
step1: Review if the validation can happen in a separate
thread
step2: Add bool canValidateAsync() const override { 
return true; 
}
In the message as public member function.
Adding a new message for Async validation require the
following extra step:
step3: Register the validated message callback in
MsgHandlersRegistrator.

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -148,13 +177,65 @@ void ReplicaImp::messageHandler(MessageBase *msg) {
       }
     }
   }
-  if (!isCollectingState() && validateMessage(trueTypeObj)) {
+
+  if (!isCollectingState()) {
+    // If replica is not doing state transfer, then for messages whose validation is needed
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:182]: Please re-phrase the comment. Not clear.

File: bftengine/src/bftengine/messages/CheckpointMsg.hpp [hunk 2]
```diff
@@ -48,6 +48,8 @@ class CheckpointMsg : public MessageBase {
 
   void validate(const ReplicasInfo& repInfo) const override;
 
+  bool canValidateAsync() const override { return true; }
```
Comment (yuliasherman) [inline bftengine/src/bftengine/messages/CheckpointMsg.hpp:51]: canValidateAsync->shouldValidateAsync

File: bftengine/src/bftengine/messages/InternalMessage.hpp [hunk 3]
```diff
@@ -63,6 +64,10 @@ using InternalMessage = std::variant<FullCommitProofMsg*,
                                      // Carries PrePrepare message after validation.
                                      PrePrepareCarrierInternalMsg,
 
+                                     // Add Carriers of messages which will translate
```
Comment (yuliasherman) [inline bftengine/src/bftengine/messages/InternalMessage.hpp:67]: It's not a translation - in fact, CarrierMesssage encapsulates an external message, right?

File: bftengine/src/bftengine/messages/InternalMessage.hpp [hunk 4]
```diff
@@ -63,6 +64,10 @@ using InternalMessage = std::variant<FullCommitProofMsg*,
...
+                                     // Incoming External messages to Internal message.
+                                     CarrierMesssage*,
```
Comment (yuliasherman) [inline bftengine/src/bftengine/messages/InternalMessage.hpp:69]: What '*' does here?

**Review by yuliasherman — APPROVED** [review]
LGTM we some comments.

File: bftengine/src/bftengine/messages/InternalMessage.hpp (hunk 3 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/messages/InternalMessage.hpp:67]: Thanks @yuliasherman , encapsulate is a better word than translate in our scenario, I will update the comment accordingly.
CarrierMessage can be thought of as a generic message which can represent a generic encapsulator message. The encapsulation (or even translation in some case) can happen for many reasons. 
In this PR we can see that the encapsulation is happening to translate an external Message into an internal message using ValidatedMessageCarrierInternalMsg.
ValidatedMessageCarrierInternalMsg<MSG> is giving us a way to do this in a generic manner so that for any new External Message and we don't have to add a new type in InternalMessage variant. All such types are represented by CarrierMessage. 
Tomorrow let's assume that we want to encapsulate for a different reason. (assume hypothetically) we want to encapsulate a request message and translate / treat the message as a reply message. We will create a RequestToReplyMsgConv<MSG> : public CarrierMessage. In that case CarrierMessage will represent either validated message or Reply Message based on the context. There can be other scenarios aswell. 
In current context as encapsulation is happening, I will update the comment.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 1 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp:182]: Thanks @yuliasherman, let me rephrase.

File: bftengine/src/bftengine/messages/InternalMessage.hpp (hunk 4 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/messages/InternalMessage.hpp:69]: CarrierMessage just know the type of the message whose encapsulation it's carrying. So I am using a Pointer Type of CarrierMessage. This means, the internal queue will point to a location which will contain an encapsulated message with the type.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 5]
```diff
@@ -69,58 +70,86 @@ using namespace concord::diagnostics;
 namespace bftEngine::impl {
 
 void ReplicaImp::registerMsgHandlers() {
-  msgHandlers_->registerMsgHandler(MsgCode::Checkpoint, bind(&ReplicaImp::messageHandler<CheckpointMsg>, this, _1));
+  msgHandlers_->registerMsgHandler(MsgCode::Checkpoint,
+                                   bind(&ReplicaImp::messageHandler<CheckpointMsg>, this, _1),
+                                   bind(&ReplicaImp::validatedMessageHandler<CheckpointMsg>, this, _1));
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaImp.cpp:75]: nit:
I assume you are registering here some messages which will never be handled by the handler?
Another option would be to have a null default value as 2nd parameter for registerMsgHandler which signals "I do not wish to register this message, it is too short to validate async..".

example:

```
msgHandlers_->registerMsgHandler(MsgCode::Checkpoint,
                                   bind(&ReplicaImp::messageHandler<CheckpointMsg>, this, _1),
                                   CheckpointMsg::shouldValidateAsync() ? 
                                   bind(&ReplicaImp::validatedMessageHandler<CheckpointMsg>, this, _1) : nullptr);
);
```

comments:

- It might look uglier, but maybe with more thinking there is a better way.
- You will need to make shouldValidateAsync static, whic hmakes sense.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 6]
```diff
@@ -148,13 +177,70 @@ void ReplicaImp::messageHandler(MessageBase *msg) {
       }
     }
   }
-  if (!isCollectingState() && validateMessage(trueTypeObj)) {
+
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaImp.cpp:180]: I think I'm confused about the above code. This is some code you had in the past specific to pre-preapre , right?
I was assuming you are planning to embed it in the common solution?
In that case, even the code handle preprepare async. , why is it left and not moved to the validation handler?

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 7]
```diff
@@ -148,13 +177,70 @@ void ReplicaImp::messageHandler(MessageBase *msg) {
...
+  // The Carrier message is allocated by this replica and after validation its given back
+  // to itself without going through the network stack (as an internal message).
+  // So the carrier message is quickly reinterpreted and used instead of fresh memory allocation.
+  // So this reinterpret_cast is intended.
+  ValidatedMessageCarrierInternalMsg<T> *validatedTrueTyeObj =
+      reinterpret_cast<ValidatedMessageCarrierInternalMsg<T> *>(msg);
+  T *trueTypeObj = validatedTrueTyeObj->returnMessageToOwner();
+  delete msg;
+
+  if (bftEngine::ControlStateManager::instance().getPruningProcessStatus()) {
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaImp.cpp:224]: can this block ever happen?
We already passed same check in ReplicaImp::messageHandler?
Maybe you should assert here?

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 8]
```diff
@@ -168,6 +254,34 @@ bool ReplicaImp::validateMessage(MessageBase *msg) {
...
+ * the validation in a thread bag and returns peacefully. This will also translate the message
+ * into internal message.
+ *
+ * @param msg : This is the message whose validation is complex and should happen asynchronously.
+ *
+ * @return : returns nothing
+ */
+template <typename MSG>
+void ReplicaImp::asyncValidateMessage(MSG *msg) {
+  RequestThreadPool::getThreadPool().async(
```
Comment (glevkovich) [inline bftengine/src/bftengine/ReplicaImp.cpp:268]: nit:
```
I don't think compiler do optimizations here. Slightly better:
```
static std::once_flag get_once;
static concord::util::ThreadPool tp;
std::call_once(get_once, [&]() { tp = RequestThreadPool::getThreadPool(); });
tp.async( ...

```

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 9]
```diff
@@ -1797,7 +1797,7 @@ async def test_install_command(self, bft_network):
     
 
     async def try_to_unwedge(self, bft_network, bft, restart, quorum=None):
-        with trio.fail_after(seconds=60):
+        with trio.fail_after(seconds=90):
```
Comment (glevkovich) [inline tests/apollo/test_skvbc_reconfiguration.py:1800]: nit: why did you have to increase it?

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 10]
```diff
@@ -582,7 +582,7 @@ async def run_replica_tls_key_exchange_cycle(self, bft_network, replicas, affect
         for rep in affected_replicas:
             reps_data[rep] = {}
             for r in replicas:
-                cert_path = os.path.join(bft_network.certdir + "/" + str(r), str(r),"server", "server.cert")
+                cert_path = os.path.join(bft_network.certdir + "/" + str(rep), str(r),"server", "server.cert")
```
Comment (glevkovich) [inline tests/apollo/test_skvbc_reconfiguration.py:585]: nit: did you find a bug here? maybe it should be in a seperate commit

File: bftengine/src/bftengine/messages/ValidatedMessageCarrierInternalMsg.hpp [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+ public:
+  CarrierMesssage(MsgType msgType) : msgType_(msgType) {}
+  virtual ~CarrierMesssage() {}
+  MsgType getMsgType() const { return msgType_; }
+
+ private:
+  MsgType msgType_;
+};
+
+// One usecase of translation of message is doing validation in a separate thread and then
```
Comment (glevkovich) [inline bftengine/src/bftengine/messages/ValidatedMessageCarrierInternalMsg.hpp:33]: typo
usecase -> use case

**Review by glevkovich — APPROVED** [review]
@arc-vmware , I think it looks really good, left a few comments and questions.
Only important thing I do not understand is why we are still processing async. the preprepare message using validatePrePrepareMsg. Is it due to the fact that this is a message that might batch other messages? Can't we have the code moved into the general case? Or is it too complex?

File: bftengine/src/bftengine/MsgHandlersRegistrator.hpp [hunk 12]
```diff
@@ -13,12 +13,14 @@
 
 #include "messages/MessageBase.hpp"
 #include "messages/InternalMessage.hpp"
+#include "messages/ValidatedMessageCarrierInternalMsg.hpp"
```
[the remark this answers is not in the document]
Comment (arc-vmware) [reply bftengine/src/bftengine/MsgHandlersRegistrator.hpp:16]: thanks @glevkovich , i will make sure that this order is maintained for all the files I touched in this PR.

File: bftengine/src/bftengine/MsgHandlersRegistrator.hpp [hunk 13]
```diff
@@ -13,12 +13,14 @@
...
 #include <functional>
 #include <unordered_map>
 
 namespace bftEngine::impl {
 
 typedef std::function<void(MessageBase*)> MsgHandlerCallback;
+typedef std::function<void(CarrierMesssage*)> ValidatedMsgHandlerCallback;
```
[the remark this answers is not in the document]
Comment (arc-vmware) [reply bftengine/src/bftengine/MsgHandlersRegistrator.hpp:23]: @glevkovich thanks, will convert this to using.

File: bftengine/src/bftengine/MsgHandlersRegistrator.hpp [hunk 14]
```diff
@@ -42,10 +59,20 @@ class MsgHandlersRegistrator {
     return nullptr;
   }
 
+  // Returns the validated callback message handler for the given message id.
+  // Validated callback is possible only for External Messages which were validated,
+  // so they are the subset of message which has a valud message type (msgId).
```
[the remark this answers is not in the document]
Comment (arc-vmware) [reply bftengine/src/bftengine/MsgHandlersRegistrator.hpp:64]: thanks @glevkovich , this is very helpful.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 5 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp:75]: Thanks @glevkovich, this is a great comment to discuss and then finalize.
Comment 1:
I thought about it, to add the handler for specific messages. But later I thought that it might increase an extra step when we would like to add a new message for asynchronous validation. If we add all message then choice of adding or not can happen through shouldValidateAsync(). So I preferred this way.
Comment 2:
Yes we can make it async, but sometimes for some complex message, it might be possible that whether to do async validation or not, will depend on some other cases. e,g, there are many kind of client req messages. Some might not be originating from clients and this may not have any signature. So we can decide case by case whether for this message any asynchronous validation is needed or not.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 6 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp:180]: There are 3 kinds of validation case and 2 kind of message handlers.
Validation case:
1) Generic sync validation.
2) Generic Async validation
3) PrePrepare specific Async validation.

When a message comes its handled by messageHandler() which decides which of the above validation will take place.
If its sync validation then the onMessage is also called.
If it's any kind of async validation, then the validated message handler will be called later. The validated message handler will have to check the pruning status and state transfer status again, before calling the onMessage.
With the above background, below is the ans to the ques:
1) This code is not for preprepare, its for all messages. This is for sync validation.
2) Currently I haven't merged the async validation, since the pre process result validation is specific to preprepare. This flow need to be completely checked, so it should be taken separately. I think its doable, and it will definitely reduce the complexity.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 7 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp:224]: Pruning happens when all nodes are ready (its not based on quorum). During pruning, it sets a replica wide flag as during pruning we should never update the blockchain. So before starting the onMessage, the handler checks pruning flag and allows only client req messages and ignores all other protocol messages. As validation would have happened in a separate thread, both pruning and state transfer need to be checked.

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 9 above)
Comment (arc-vmware) [reply tests/apollo/test_skvbc_reconfiguration.py:1800]: This will be reverted as all the reconfiguration testcases are fixed in another PR. Once that goes in I will merge and commit.

File: tests/apollo/test_skvbc_reconfiguration.py (hunk 10 above)
Comment (arc-vmware) [reply tests/apollo/test_skvbc_reconfiguration.py:585]: This will also be reverted.

File: bftengine/src/bftengine/messages/ValidatedMessageCarrierInternalMsg.hpp (hunk 11 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/messages/ValidatedMessageCarrierInternalMsg.hpp:33]: Thanks I will change this.

**Comment by arc-vmware** [comment]
@glevkovich Thanks a lot for the review.
The preprepare message is a set of many different kind of client req message. There is a specific validation that is happening within a subset of the the set of client message. This validation requires all the messages of subset to finalize the final verdict of the validation. The name of the message is pre process result validation. This validation is done in replicaimp and if validation fails then an instant view change is needed.
To change this, we have to move this to Prepreparemsg's validation and then pursue the natural course of view change. 
As this is more related to complexity reduction and less with performance, I am thinking of taking this in a separate PR. For now I will raise a JIRA as this is also an important task.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 8 above)
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp:268]: Thanks @glevkovich , this is good. As the RequestThreadPool::getThreadPool() returns a reference to the static threadpool, we directly use the same. This is good for low level, and it definitely improves the understanding.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 15]
```diff
@@ -1323,13 +1437,40 @@ void ReplicaImp::onMessage<FullCommitProofMsg>(FullCommitProofMsg *msg) {
...
+ */
+void ReplicaImp::onCarrierMessage(CarrierMesssage *msg) {
+  auto validatedMsgCallBack = msgHandlers_->getValidatedMsgCallback(msg->getMsgType());
+  if (validatedMsgCallBack) {
+    validatedMsgCallBack(msg);
+  } else {
+    // Below delete can be a memory leak and must be investigated.
+    // This can happen only when non-translatable message is sent,
+    // this should never happen.
+    LOG_ERROR(GL, "Received unknown external translated Message: " << KVLOG(msg->getMsgType()));
```
[the remark this answers is not in the document]
Comment (arc-vmware) [reply bftengine/src/bftengine/ReplicaImp.cpp:1456]: Thanks @glevkovich , let me use concord assert here.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 6 above)
Comment (glevkovich) [reply bftengine/src/bftengine/ReplicaImp.cpp:181]: @arc-vmware 
OK thanks for the explenation - answer 1 - I compeltely understood, for answer 2 - I now I understand you leave the case of merging the cases for the future.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 8 above)
Comment (glevkovich) [reply bftengine/src/bftengine/ReplicaImp.cpp:269]: @arc-vmware , yes - my suggestion was on how to improve the fetch of this reference (fetch only once and not millions of times), not on it's creation which is done only once.
